README.md
# HEPTAconnect Portal Local Shopware Platform 
#### This is part of HEPTACOM solutions for medium and large enterprises.

## Description

This is a HEPTAconnect package that allows to communicate multiple entity types with a Shopware 6 instance that also integrates HEPTAconnect.
Read more in the [documentation](https://heptaconnect.io/).


## System requirements

* PHP 7.4 or above


## Changelog

See the attached [CHANGELOG.md](./CHANGELOG.md) file for a complete version history and release notes.


## Additional development requirements

* Make
* Any debugging/coverage php extension like xdebug or pcov


## Contributing

Thank you for considering contributing to this package! Be sure to sign the [CLA](./CLA.md) after creating the pull request. [![CLA assistant](https://cla-assistant.io/readme/badge/HEPTACOM/heptaconnect-portal-local-shopware-platform)](https://cla-assistant.io/HEPTACOM/heptaconnect-portal-local-shopware-platform)


### Steps to contribute

1. Fork the repository
2. `git clone yourname/heptaconnect-portal-local-shopware-platform`
3. Make your changes to master branch
4. Create your Pull-Request


### Check your changes

1. Check and fix code style `make cs-fix && make cs`
2. Check tests `make test`
3. Check whether test code coverage is same or higher `make coverage`
4. Check whether tests can find future obscurities `make infection`


## License

Copyright 2020 HEPTACOM GmbH

Dual licensed under the [GNU Affero General Public License v3.0](./LICENSE.md) (the "License") and proprietary license; you may not use this project except in compliance with the License.
You may obtain a copy of the AGPL License at [https://spdx.org/licenses/AGPL-3.0-or-later.html](https://spdx.org/licenses/AGPL-3.0-or-later.html).
Contact us on [our website](https://www.heptacom.de) for further information about proprietary usage.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to a variation of [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
The version numbers are structured like `GENERATION.MAJOR.MINOR.PATCH`:

* `GENERATION` version when concepts and APIs are abandoned, but brand and project name stay the same,
* `MAJOR` version when you make incompatible API changes and provide an upgrade path,
* `MINOR` version when you add functionality in a backwards compatible manner, and
* `PATCH` version when you make backwards compatible bug fixes.

## [Unreleased]

### Added

- Add guessing of mimeType and extension in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\MediaUnpacker`
- Add composer dependency `ext-fileinfo:*` for mimeType guessing in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\MediaUnpacker`
- Add `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Packer\OrderCustomerPacker` to pack an order customer into a customer dataset entity
- Add update logic for customer addresses in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\CustomerReceiver`

### Changed

### Deprecated

### Removed

### Fixed

- Add array_filter in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\TranslatableUnpacker` to remove null values from translation locale result to prevent array offset exception.
- Fix a bug when mapping property ids with a composite primary key in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\ProductReceiver`
- Add try catch on product unpacking in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\ProductReceiver` to prevent a complete fail of the whole batch.
- Fix conversion of Shopware product lineitems depending on the products to still exist in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Emitter\OrderEmitter`
- Fix emission of Shopware order, when a customer account is not present anymore, by relying on information of the order customer
- Use same salutation lookup for customer address and customer in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\CustomerReceiver`, so customer do not only use fallback salutation
- Fix order emission by checking if an order-transaction exists before usage in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Emitter\OrderEmitter`
- Fix order reception by adding the product number to newly created line items in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\OrderReceiver`

### Security

## [0.9.0.2] - 2022-06-01

### Fixed

- Fixed address splitting on customer creation in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Receiver\CustomerReceiver::getAddress` to satisfy the shopware standard address representation regarding street and house number. 

## [0.9.0.1] - 2022-04-23

### Fixed

- Fix portal node service container extension to work with both portal FQCNs `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\LocalShopwarePlatformPortal` and the deprecated `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Portal`

## [0.9.0.0] - 2022-04-05

### Added

- Interpret `\Heptacom\HeptaConnect\Dataset\Ecommerce\Product\Product::getMedias` to receive medias on products

### Changed

- Use `\Heptacom\HeptaConnect\Dataset\Ecommerce\Media\Media::getFile` for transferring instead of `\Heptacom\HeptaConnect\Dataset\Ecommerce\Media\Media::getNormalizedStream`

### Deprecated

- Deprecate `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Portal` as renamed to `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\LocalShopwarePlatformPortal`

### Removed

- Remove support for `shopware/core: >=6.2.1 <6.4`

### Fixed

- Product medias unpacked by `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\ProductUnpacker` have position by appearance in the product entity

## [0.8.2] - 2022-02-09

### Fixed

- Fix function call on null in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Emitter\OrderEmitter`.
- Fix compatibility with `shopware/core:^6.4` by supporting `\Shopware\Core\System\Currency\CurrencyEntity::setItemRounding` if it exists.

## [0.8.1] - 2021-12-07

### Fixed

- Fetch VAT-ID from customer instead of address in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Packer\CustomerPacker::pack`

## [0.8.0] - 2021-11-22

### Added

- Add `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support\LocaleMatcher` to centralize translation handling of incoming locale matching
- Add `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\TranslatableUnpacker` to centralize translations payload generation of any `\Heptacom\HeptaConnect\Dataset\Base\Translatable\Contract\TranslatableInterface`
- Add log message code `1637342440` in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support\LocaleMatcher::match` when a locale code is tested against a Shopware language
- Add log message code `1637342441` in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support\LocaleMatcher::match` when a locale code could not be matched
- Add log message code `1637342442` in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support\LocaleMatcher::match` when a locale code could be matched to a unique Shopware language
- Add log message code `1637342443` in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support\LocaleMatcher::match` when a locale code could be matched against multiple other Shopware languages
- Add log message code `1637344184` in `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\TranslatableUnpacker::unpack` when a translated value is tried to be applied but the language code could not be mapped to a Shopware language
- Add `\Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Unpacker\CategoryUnpacker` to unpack category data into Shopware API payload

src/Support/LocaleMatcher.php
<?php

declare(strict_types=1);

namespace Heptacom\HeptaConnect\Portal\LocalShopwarePlatform\Support;

use Psr\Log\LoggerInterface;

class LocaleMatcher
{
    private LoggerInterface $logger;

    private int $matchDistance;

    public function __construct(LoggerInterface $logger, int $matchDistance = 8)
    {
        $this->logger = $logger;
        $this->matchDistance = $matchDistance;
    }

    public function match(array $haystack, string $needle): ?string
    {
        if (\in_array($needle, $haystack, true)) {
            return $needle;
        }

        if ($haystack === []) {
            return null;
        }

        $matches = [];

        foreach ($haystack as $straw) {
            $distance = \levenshtein($needle, $straw, 1, 4, 10);
            $this->logger->debug('LocaleMatcher: Language match test result', [
                'test' => $needle,
                'comparison' => $straw,
                'distance' => $distance,
                'considered' => $distance <= $this->matchDistance,
                'code' => 1637342440,
            ]);

            if ($distance <= $this->matchDistance) {
                $matches[] = $straw;
            }
        }

        if ($matches === []) {
            $this->logger->error('LocaleMatcher: Could not find a matching known locale code', [
                'locale' => $needle,
                'options' => $haystack,
                'code' => 1637342441,
            ]);

            return null;
        }

        $result = \array_shift($matches);

        if ($matches === []) {
            $this->logger->info('LocaleMatcher: Matching known locale code', [
                'locale' => $needle,
                'options' => $haystack,
                'result' => $result,
                'code' => 1637342442,
            ]);
        } else {
            $this->logger->warning('LocaleMatcher: Found too many matching locale codes', [
                'locale' => $needle,
                'options' => $haystack,
                'result' => $result,
                'other_matches' => $matches,
                'code' => 1637342443,
            ]);
        }

        return $result;
    }
}
